### ISSUE TITLE
No suitable candidates found

### ISSUE THREAD

**Issue by enterux** [richardscreene/mod_janus#1]
**Describe the bug**
Try to join SIP Call to Janus Room, the plugin joins the user, however the media is not being negotiated properly.
Always end up with a NETWORK_OUT_OF_ORDER 

2020-09-19 23:39:18.888526 [DEBUG] switch_core_media.c:5594 Audio Codec Compare [opus:102:48000:20:0:1]/[opus:116:48000:20:0:1]
2020-09-19 23:39:18.888526 [DEBUG] switch_core_media.c:5649 Audio Codec Compare [opus:116:48000:20:0:1] ++++ is saved as a match
2020-09-19 23:39:18.888526 [DEBUG] mod_opus.c:608 Opus encoder: set bitrate based on maxaveragebitrate value found in SDP or local config [64000bps]
2020-09-19 23:39:18.888526 [DEBUG] mod_opus.c:624 Opus encoder: set audio bandwidth to [FULLBAND] based on maxplaybackrate value found in SDP or local config [48000Hz]
2020-09-19 23:39:18.888526 [DEBUG] mod_opus.c:608 Opus encoder: set bitrate based on maxaveragebitrate value found in SDP or local config [64000bps]
2020-09-19 23:39:18.888526 [DEBUG] mod_opus.c:624 Opus encoder: set audio bandwidth to [FULLBAND] based on maxplaybackrate value found in SDP or local config [48000Hz]
2020-09-19 23:39:18.888526 [DEBUG] switch_core_media.c:3839 Set Codec janus/localbox/7021477573@2422 opus/48000 20 ms 960 samples 0 bits 1 channels
2020-09-19 23:39:18.888526 [DEBUG] switch_core_codec.c:111 janus/localbox/7021477573@2422 Original read codec set to opus:116
2020-09-19 23:39:18.888526 [DEBUG] switch_core_media.c:4329 Searching for rtp candidate.
2020-09-19 23:39:18.888526 [DEBUG] switch_core_media.c:4329 Searching for rtcp candidate.
2020-09-19 23:39:18.888526 [DEBUG] switch_core_media.c:4376 Look for Relay Candidates as last resort
2020-09-19 23:39:18.888526 [DEBUG] switch_core_media.c:4329 Searching for rtp candidate.
2020-09-19 23:39:18.888526 [DEBUG] switch_core_media.c:4329 Searching for rtcp candidate.
2020-09-19 23:39:18.888526 [DEBUG] switch_core_media.c:4382 janus/localbox/7021477573@2422 no suitable candidates found.
2020-09-19 23:39:18.888526 [DEBUG] switch_core_media.c:5882 No 2833 in SDP. Liberal DTMF mode adding 0 as telephone-event.
2020-09-19 23:39:18.888526 [DEBUG] switch_core_media.c:5918 No 2833 in SDP.  Disable 2833 dtmf and switch to INFO
2020-09-19 23:39:18.888526 [INFO] mod_janus.c:193 Cannot negotiate SDP

janus.conf.xml

<configuration name="janus.conf" description="Janus Endpoint">
  <settings>
    <param name="debug" value="true"/>
  </settings>

  <server name="localbox">
    <param name="url" value="http://localhost:8088/janus"/>
    <param name="secret" value="somesecret"/>
    <param name="auth-token" value="somesecret"/>
    <param name="enabled" value="true"/>
    <param name="rtp-ip" value="$${bind_server_ip}"/>
    <param name="apply-candidate-acl" value="localnet.auto"/>
    <param name="apply-candidate-acl" value="rfc1918.auto"/>
    <param name="local-network-acl" value="localnet.auto"/>
    <param name="apply-candidate-acl" value="wan.auto"/>
    <param name="ext-rtp-ip" value="auto-nat"/>
    <param name="codec-string" value="opus"/>
  </server>
</configuration>

Dialplan

 <extension name="public_did">
    <condition field="destination_number" expression="^7193$">
        <action application="answer" data=""/>
        <action application="log" data="INFO Joining Call to Conference"/>
        <action application="bridge" data="{janus-use-existing-room=false,janus-start-muted=false}janus/localbox/${caller_id_number}@2422"/>
    </condition>
  </extension>

**To Reproduce**
1. Compile latest freeswitch from github repo
2.  Compile mod_janus along with it using the Readme.md method
3. Created dialplan to add janus http API with the right apisecret 
4. Call the extension and it always fails with the above error.

**Expected behavior**
Join the AudioBridge Room 2422 which is already active and has other participants via browser.

**Versions (please complete the following information):**
 - FS Version : FreeSWITCH Version 1.10.6-dev+git~20200916T141527Z~89c7f88656~64bit (git 89c7f88 2020-09-16 14:15:27Z 64bit)
 - Janus Version : v0.10.4

**Comment by richardscreene** [comment]
As you say it looks like Janus isn't returning any suitable candidates.  Get a freeswitch sip trace so we can prove this 
I would guess that Janus isn't configured correctly.  Have you defined a turn server?

**Comment by enterux** [comment]
Yes Janus has Turn server configured.

But it seems that Janus sends back SDP which doesnt have any candidates for RTP Transmission to establish.
https://pastebin.com/FwpGGHWV
Much better to read :)

Both Janus and FreeSWITCH are on same physical box.
Chrome browser is able to join the AudioBridge perfectly well, so TURN is working IMO
However the clients trying to join via FS are not able to establish the connection due to RTP failure.

**Comment by enterux** [comment]
Both Janus and FreeSWITCH are running on the same server.

**Comment by richardscreene** [comment]
Ah, looks like mod_janus doesn't handle trickle sending of candidates from Janus.
It's the first time I've seen Janus send trickle candidates!   Can you configure Janus to send candidates in the offer?

I will add trickle functionality when I get some time.

**Comment by enterux** [comment]
I am not sure how to configure Janus to send candidates right in the offer. But my guess is, it does send in offer, i think this is default functionality of janus to send trickle candidates based on what it understands about the client state change via TURNSERVER.

**Comment by enterux** [comment]
Ok update here : 
I set full_trickle=false in Janus and the call got accepted.

So It works, however, adding trickle functionality in mod_janus should be handy.

Let me know,

**Comment by richardscreene** [comment]
I agree.  I'll leave this issue open until I have time to look into it.  In the meantime I'll update the reader.
Thanks for reporting!!!

**Comment by richardscreene** [comment]
BTW, the correct repo for mod_janus is https://github.com/freeswitch/mod_janus
There is no guarantee that this repo will be kept up to date.

**Comment by enterux** [comment]
Yeah, I used this repo in the FreeSWITCH compilation. Included this in modules.conf

**Comment by richardscreene** [comment]
OK, I'm going to disable this repo.  All activity should be on the freeswitch/mod_janus repo
